✨amp-experiment 1.0: Created the Base Allow List for Attribute Mutations

torch2424 opened the pull request:
relates to #20225
relates to #21705

This creates the base allow list for attribute mutations. **NOTE:** This should not break the current 1.0 functionality, as all tests are passing and everything works as it should! 😄 

**Tests will come after the initial review, but are already completed** 👍 

# Example

<img width="1440" alt="Screen Shot 2019-05-16 at 2 10 30 PM" src="https://user-images.githubusercontent.com/1448289/57896808-6bc9c400-7807-11e9-94c8-207ffec0cc24.png">

torch2424 commented:
@zhouyx Made your requested offline changes (some naming things, and slight logic modifications), this should be good to go 😄

zhouyx commented on extensions/amp-experiment/1.0/mutation-parser.js:
```diff
@@ -65,23 +36,35 @@ const SUPPORTED_ATTRIBUTES = {
-    assertCharacterDataMutation(mutationRecord, stringifiedMutation);
+    assertCharacterDataMutationFormat(mutationRecord, stringifiedMutation);
 
     return () => {
```
We've talked about this offline. If we are going to add `innerHTML` and `reorder` mutation, I think we will need to unify the format. Maybe create an interface or class for mutationEntry that `allowedAttributeMutationEntry` implement. 
We don't need to worry about that yet, but could you please leave a comment that this is a mutationEntry where validate() is noop.

zhouyx commented on extensions/amp-experiment/1.0/mutation-parser.js:
```diff
@@ -65,23 +36,35 @@ const SUPPORTED_ATTRIBUTES = {
 
   const stringifiedMutation = JSON.stringify(mutation);
 
   setSelectorToElement('target', mutationRecord, document);
```
If the first argument is always going to be 'target', can we get rid of it?

zhouyx commented on extensions/amp-experiment/1.0/attribute-allow-list/allowed-attribute-mutation-entry.js:
```diff
@@ -0,0 +1,114 @@
+   * @param {!Object} mutationRecord
+   */
+  mutate(mutationRecord) {
+    mutationRecord['targetElement'].setAttribute(
```
I couldn't find it, but could you point me to where `mutationRecord['targetElement']` is defined? Thank you

torch2424 replied:
So where you commneted: "If the first argument is always going to be 'target', can we get rid of it?"

`setSelectorToElement()` is what sets the `targetElement` 😄

zhouyx commented on extensions/amp-experiment/1.0/attribute-allow-list/allowed-attribute-mutation-entry.js:
```diff
@@ -0,0 +1,114 @@
+
+  /** @override */
+  validate(value) {
+    assertHttpsUrl(value, 'attributes', 'mutation');
```
Hmm looking at the `assertHttpsUrl` definition https://github.com/ampproject/amphtml/blob/master/src/url.js#L306. I don't think `attributes` and `mutation` are the two arguments that we want to pass in. Could you double check. Thank you

zhouyx commented on extensions/amp-experiment/1.0/attribute-allow-list/attribute-allow-list.js:
```diff
@@ -0,0 +1,86 @@
+
+  // Search through the allow list for our
+  // Allowed attribute entry
+  let allowedAttributeMutationEntry = undefined;
```
nit: no need to specify undefined : ) it's by default initialized to undefined.

zhouyx commented on extensions/amp-experiment/1.0/attribute-allow-list/attribute-allow-list.js:
```diff
@@ -0,0 +1,86 @@
+  // Search through the allow list for our
+  // Allowed attribute entry
+  let allowedAttributeMutationEntry = undefined;
+  if (attributeMutationAllowList[mutationAttributeName][mutationTagName]) {
```
I'm a bit concerned about the number of entries of the `attributeMutationAllowList[mutationAttributeName]` if there're many `mutationTagNames` that we support. For example `src` maybe allowed for a list of tags, are we going to have an instance for each of them.

torch2424 replied:
Yes, though the list of `src` supported elements is small: https://www.w3schools.com/tags/att_src.asp

Src is an interesting case though, as for not, it should only be amp-img.

In my opinion, even in the worst case scenario of everything being done on the client, I don't think this list would get extremely large if we do it an on a request basis. Especially since we are implementing '*' values for tag names.

Let me know if this makes sense.  I feel that is more of a design questions. In my opinion, Allow list offer security and verbosity, at the expense of being so verbose. We can take this offline if you think it helps 👍

zhouyx commented on extensions/amp-experiment/1.0/attribute-allow-list/attribute-allow-list.js:
```diff
@@ -0,0 +1,86 @@
+    allowedAttributeMutationEntry =
+      attributeMutationAllowList[mutationAttributeName]['*'];
+
+    if (!allowedAttributeMutationEntry.tags) {
```
Looked at the initialization below. I found that the tags is undefined for '*'.  Which I don't think is what we want here? Am I missing anything.

torch2424 replied:
For Default Style, and Default URL, the base `tags` is `['*']`.

It's defined in the class constructor. 😄 

And I'm, using the if block, and than an else if, since the if would be execute first, then the next condition 👍

zhouyx replied:
ping on the question : ) And I see your test case. 
When would we have
`'*': new FakeAllowedAttributeMutationEntry('tagName')`?

zhouyx commented on extensions/amp-experiment/1.0/attribute-allow-list/attribute-allow-list.js:
```diff
@@ -0,0 +1,86 @@
+    '*': new DefaultStyleAllowedAttributeMutationEntry(),
+  },
+  'src': {
+    '*': new DefaultAllowedURLAttributeMutationEntry(),
```
To help me better understand. Could you give me an example on what the list would look like with tagName `amp-img` and `amp-iframe`? Thank you

torch2424 replied:
Oh yeah totally:

```
export const attributeMutationAllowList = {
  'style': {
    '*': new DefaultStyleAllowedAttributeMutationEntry(),
  },
  'src': {
    'amp-img': new AmpBindURLAttributeMutationEntry(),
    'amp-iframe': new AMPBindURLAttributeMutationEntry(),
    '*': new DefaultAllowedURLAttributeMutationEntry(),
  },
  'href': {
    '*': new DefaultAllowedURLAttributeMutationEntry(),
  },
};
```

where the example: `AmpBindURLAttributeMutationEntry` extends the `DefaultAllowedURLAttributeMutationEntry`, and simply changes the mutate function  😄

torch2424 commented:
@zhouyx Replied to comments / made changes. Let me know what you think! 😄

torch2424 commented:
@zhouyx 

> I was imaging something like 'amp-img|amp-iframe' : AMPBindURLAttributeMutationEntry(), but that add look up time a lot. So maybe not... Any other ideas?

Yeah I agree that would increase look up time, which I'm trying to avoid. Personally, I can't think of something else, but perhaps we can whiteboard something? That being said, I don't think two entries sharing the same entry should be too bad.I feel like we will either be duplicating entries, or duplicating tags if we go the other way around. Though, the latter would have slower lookup times. I'm open to designing something on this, though, I feel like there will be some kind of tradeoff at the end of the day.

And with that, I have added the tests, fixed comments, and replied to them.

This is good to go! 😄

zhouyx commented on extensions/amp-experiment/1.0/test/test-attribute-allow-list.js:
```diff
@@ -0,0 +1,152 @@
+
+  it('should allow a valid tag, attribute, value', () => {
+    expect(
+      getAllowedAttributeMutationEntry(getAttributeMutation(), MUTATION_NAME)
```
nit: is it possible to do `to.be.instanceOf(DefaultAllowedURLAttributeMutationEntry)`?

zhouyx replied:
instead of `MUTATION_NAME`, can we do `JSON.stringify(getAttributeMutation()`? That's how the stringified mutation is calculated. 
Or are you using `MUTATION_NAME` for simplicity?

torch2424 replied:
Did the nit, and the second is yes, it's for simplicity 😄

zhouyx commented on extensions/amp-experiment/1.0/test/test-attribute-allow-list.js:
```diff
@@ -0,0 +1,152 @@
+  });
+
+  it('should travese by attributeName', () => {
+    const testAttributeName = 'test';
```
I just realize we convert all tagName to lowerCase(). What about we put `TEST: new AllowedAttributeMutationEntry()` instead, so we can avoid the `toLowerCase()` conversion. But it's minor, and completely up to you : )

zhouyx replied:
Do we have to use 'test', can we use the supported one 'style' instead?

zhouyx commented on extensions/amp-experiment/1.0/test/test-attribute-allow-list.js:
```diff
@@ -0,0 +1,152 @@
+        getAttributeMutation(testAttributeName),
+        MUTATION_NAME
+      )
+    ).to.be.ok;
```
Same, can we create a fake MutationEntry for testing here, put 
```
attributeMutationAllowList[testAttributeName] = {
      '*': new FakeAllowedAttributeMutationEntry(['*']),
    };
```
And then check that the returned value is an instance of `FakeAllowedAttributeMutationEntry`?

zhouyx commented on extensions/amp-experiment/1.0/test/test-attribute-allow-list.js:
```diff
@@ -0,0 +1,152 @@
+        getAttributeMutation(testAttributeName),
+        MUTATION_NAME
+      )
+    ).to.be.ok;
```
I found the test to not be sufficient. It's difficult to tell if the mutationEntry is for the 'div' only or it works for all tagName. Would be great to test with a different tag combined.

zhouyx commented on extensions/amp-experiment/1.0/test/test-attribute-allow-list.js:
```diff
@@ -0,0 +1,152 @@
+  it('should handle * tagName, and use specifc tag in entry tags array', () => {
+    const testAttributeName = 'test';
+    attributeMutationAllowList[testAttributeName] = {
+      '*': new AllowedAttributeMutationEntry(['div']),
```
Will this ever happen? `*` with `div`

zhouyx commented on extensions/amp-experiment/1.0/test/test-attribute-allow-list.js:
```diff
@@ -0,0 +1,152 @@
+  });
+
+  it('should travese by tagName', () => {
+    const testAttributeName = 'test';
```
I understand you want to test the `getAttributeMutation()` function, but not the real list. But I personally think it'd be great if we could test with some real cases once we have `amp-img` maybe.

torch2424 commented:
Talked offline, also agreed to remove `whateverAttributeMutationEntry.tags`, to make things simpler. Can be added later if the list gets too large 😄

torch2424 commented:
@zhouyx Made requested changes this is good to go 😄

zhouyx commented on extensions/amp-experiment/1.0/attribute-allow-list/allowed-attribute-mutation-entry.js:
```diff
@@ -0,0 +1,95 @@
+   * @param {!Object} mutationRecord
+   * @return {boolean}
+   */
+  validate(mutationRecord) {
```
Is the `AllowedAttributeMutationEntry` base class's `#validate ()` method used by any extended class? 
Given there're only two methods, and each extended class is likely going to implement its own version of the `#validate()` function. I start to feel that using interface class, and have each class implement it might be a better solution compare to extend on top of base. 

I don't want to block your following PR. I'll leave the decision to you : )

torch2424 replied:
Accidentally clicked this when working on side projects. So I'm replying so I remember Tuesday :joy: 

But TL;DR Yup, class mutations will use validate. But I'm okay with an interface :+1: Let's chat Tuesday as I can explain how some override on or the other :smile:

torch2424 commented:
Talked offline with @zhouyx Agreed to merge this PR and the next. And move over everything to interface, since we will then have the base in place. 😄
